In [1]:
from utils import max_features, maxlen, embed_size_fastText, embed_size_glove, embed_size_glove_twitter
from utils import batch_size, epochs, run_5fold, write_prediction, eval_results
from model import get_cnn_inception
import pickle
import tensorflow as tf
import pandas as pd

/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
Using TensorFlow backend.
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #2
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/matplotlib/__init__.py:962: UserWarning: Duplicate key in file "/home/ec2-user/.config/matplotlib/matplotlibrc", line #3
  (fname, cnt))
/home/ec2-user/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/sklearn/cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into whic

In [2]:
fastText = pickle.load(open("input/fastText.300d.pickle", "rb"))
glove = pickle.load(open("input/glove.300d.pickle", "rb"))
glove_twitter = pickle.load(open("input/glove.twitter.200d.pickle", "rb"))

In [3]:
x_train = pickle.load(open("input/x_train.pickle", "rb"))
y_train = pickle.load(open("input/y_train.pickle", "rb"))
x_test = pickle.load(open("input/x_test.pickle", "rb"))
submission = pd.read_csv("input/sample_submission.csv")

## check different embedding

In [4]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "num_filter": 50,
    "reg": 0.0,
    "dropout": 0.2
}

In [5]:
embeddings = [(fastText, embed_size_fastText, "fastText"), (glove, embed_size_glove, "glove"), (glove_twitter, embed_size_glove_twitter, "glove_twitter")]

In [6]:
predictions = {}

In [7]:
for embed in embeddings:
    print ("\n********************{}********************".format(embed[2]))
    params["embedding_matrix"] = embed[0]
    params["embed_size"] = embed[1]
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_cnn_inception, **params)
    predictions[embed[2]] = y_preds


********************fastText********************
Instructions for updating:
Please switch to tf.metrics.auc. Note that the order of the labels and predictions arguments has been switched.
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 106s - loss: 0.0702 - acc: 0.9772 - auc_roc: 0.9086 - val_loss: 0.0465 - val_acc: 0.9826 - val_auc_roc: 0.9621
Epoch 2/50
 - 103s - loss: 0.0444 - acc: 0.9833 - auc_roc: 0.9710 - val_loss: 0.0444 - val_acc: 0.9828 - val_auc_roc: 0.9756
Epoch 3/50
 - 103s - loss: 0.0382 - acc: 0.9853 - auc_roc: 0.9787 - val_loss: 0.0455 - val_acc: 0.9827 - val_auc_roc: 0.9808
Epoch 00003: early stopping
val loss: 0.04554953154724499
val acc: 0.9826674824446024
val auc roc: 0.9808075894462303

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 106s - loss: 0.0795 - acc: 0.9758 - auc_roc: 0.9028 - val_loss: 0.0465 - val_acc: 0.9824 - val_auc_roc: 0.9617
Epoch 2/50
 - 103s - loss: 0.0446 - acc: 0.9832 - auc_roc: 0.9704 - val_loss: 0.0442 - val

Epoch 4/50
 - 93s - loss: 0.0378 - acc: 0.9854 - auc_roc: 0.9801 - val_loss: 0.0463 - val_acc: 0.9832 - val_auc_roc: 0.9813
Epoch 00004: early stopping
val loss: 0.04633463971737506
val acc: 0.9832205085622313
val auc roc: 0.9813034348621729

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 95s - loss: 0.0791 - acc: 0.9759 - auc_roc: 0.9014 - val_loss: 0.0482 - val_acc: 0.9824 - val_auc_roc: 0.9587
Epoch 2/50
 - 93s - loss: 0.0467 - acc: 0.9826 - auc_roc: 0.9679 - val_loss: 0.0458 - val_acc: 0.9829 - val_auc_roc: 0.9727
Epoch 3/50
 - 93s - loss: 0.0416 - acc: 0.9841 - auc_roc: 0.9759 - val_loss: 0.0455 - val_acc: 0.9832 - val_auc_roc: 0.9780
Epoch 4/50
 - 93s - loss: 0.0377 - acc: 0.9853 - auc_roc: 0.9798 - val_loss: 0.0453 - val_acc: 0.9832 - val_auc_roc: 0.9811
Epoch 5/50
 - 93s - loss: 0.0343 - acc: 0.9867 - auc_roc: 0.9822 - val_loss: 0.0468 - val_acc: 0.9834 - val_auc_roc: 0.9831
Epoch 00005: early stopping
val loss: 0.04675493347555718
val acc: 0.9833615153379859


In [8]:
eval_results(predictions)

key: fastText
loss: 0.04554556013987972 (+/- 0.0007409549985705233)
acc: 0.9828028274643259 (+/- 0.00036012916234944676)
roc_auc: 0.9803595673865427 (+/- 0.0005286042761330495)



key: glove
loss: 0.04468144988842106 (+/- 0.0007784231426185782)
acc: 0.9831161680493627 (+/- 0.0002944974584065599)
roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)



key: glove_twitter
loss: 0.04602299235323956 (+/- 0.0009691474923356518)
acc: 0.9830085934930389 (+/- 0.0005089609912555852)
roc_auc: 0.9803116532286843 (+/- 0.0028465203093875034)





##### result: glove
- fastText: 

loss: 0.04554556013987972 (+/- 0.0007409549985705233)

acc: 0.9828028274643259 (+/- 0.00036012916234944676)

roc_auc: 0.9803595673865427 (+/- 0.0005286042761330495)

- glove: 

loss: 0.04468144988842106 (+/- 0.0007784231426185782)

acc: 0.9831161680493627 (+/- 0.0002944974584065599)

roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)

- glove.twitter: 

loss: 0.04602299235323956 (+/- 0.0009691474923356518)

acc: 0.9830085934930389 (+/- 0.0005089609912555852)

roc_auc: 0.9803116532286843 (+/- 0.0028465203093875034)



In [9]:
opt = predictions['glove']

## check num_filter
(suggest values: 100~600: test 100, 300, 500)

In [10]:
filters = [100, 300]  # for the result of 50, refer to the previous log

In [11]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "reg": 0.0,
    "dropout": 0.2,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove
}

In [12]:
for f in filters:
    print ("\n********************f = {}********************".format(f))
    params["num_filter"] = f
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_cnn_inception, **params)
    predictions[f] = y_preds


********************f = 100********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 187s - loss: 0.0602 - acc: 0.9794 - auc_roc: 0.9358 - val_loss: 0.0499 - val_acc: 0.9811 - val_auc_roc: 0.9709
Epoch 2/50
 - 185s - loss: 0.0438 - acc: 0.9834 - auc_roc: 0.9764 - val_loss: 0.0446 - val_acc: 0.9829 - val_auc_roc: 0.9796
Epoch 3/50
 - 185s - loss: 0.0387 - acc: 0.9851 - auc_roc: 0.9818 - val_loss: 0.0440 - val_acc: 0.9831 - val_auc_roc: 0.9832
Epoch 4/50
 - 185s - loss: 0.0343 - acc: 0.9867 - auc_roc: 0.9845 - val_loss: 0.0478 - val_acc: 0.9822 - val_auc_roc: 0.9854
Epoch 00004: early stopping
val loss: 0.04776315323678132
val acc: 0.9821713709185046
val auc roc: 0.9853609959701564

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 187s - loss: 0.0600 - acc: 0.9795 - auc_roc: 0.9411 - val_loss: 0.0456 - val_acc: 0.9827 - val_auc_roc: 0.9717
Epoch 2/50
 - 185s - loss: 0.0438 - acc: 0.9834 - auc_roc: 0.9770 - val_loss: 0.0456 - val_acc: 0.9825 - 

KeyboardInterrupt: 

In [13]:
eval_results(predictions)

key: fastText
loss: 0.04554556013987972 (+/- 0.0007409549985705233)
acc: 0.9828028274643259 (+/- 0.00036012916234944676)
roc_auc: 0.9803595673865427 (+/- 0.0005286042761330495)



key: glove
loss: 0.04468144988842106 (+/- 0.0007784231426185782)
acc: 0.9831161680493627 (+/- 0.0002944974584065599)
roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)



key: glove_twitter
loss: 0.04602299235323956 (+/- 0.0009691474923356518)
acc: 0.9830085934930389 (+/- 0.0005089609912555852)
roc_auc: 0.9803116532286843 (+/- 0.0028465203093875034)



key: 100
loss: 0.045756475632165755 (+/- 0.0011660776254660248)
acc: 0.9826607834588188 (+/- 0.00033719530397060746)
roc_auc: 0.9828798976335431 (+/- 0.0017559481870277032)



key: 300
loss: 0.04898223692728457 (+/- 0.0024399659468127824)
acc: 0.9822043513803225 (+/- 0.0006823144195393634)
roc_auc: 0.9806345902075753 (+/- 0.0016923229137481245)





###### result: 

- f = 50

loss: 0.04468144988842106 (+/- 0.0007784231426185782)

acc: 0.9831161680493627 (+/- 0.0002944974584065599)

roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)


- f = 100

loss: 0.045756475632165755 (+/- 0.0011660776254660248)

acc: 0.9826607834588188 (+/- 0.00033719530397060746)

roc_auc: 0.9828798976335431 (+/- 0.0017559481870277032)


- f = 300

loss: 0.04898223692728457 (+/- 0.0024399659468127824)

acc: 0.9822043513803225 (+/- 0.0006823144195393634)

roc_auc: 0.9806345902075753 (+/- 0.0016923229137481245)


In [14]:
opt = opt

## check drop out
(suggest value 0.1~0.5: test 0.1, 0.2, 0.5)

In [15]:
params = {
    "maxlen": maxlen,
    "max_features": max_features,
    "reg": 0.0,
    "embedding_matrix": glove,
    "embed_size": embed_size_glove,
    "num_filter": 50
}

In [16]:
dropouts = [0.1, 0.5] # 0.2 see previous result

In [17]:
for d in dropouts:
    print ("\n********************d = {}********************".format(d))
    params["dropout"] = d
    y_preds = run_5fold(x_train, y_train, x_test, "loss", get_cnn_inception, **params)
    predictions[d] = y_preds


********************d = 0.1********************
Train on 127656 samples, validate on 31915 samples
Epoch 1/50
 - 105s - loss: 0.0635 - acc: 0.9789 - auc_roc: 0.9281 - val_loss: 0.0463 - val_acc: 0.9827 - val_auc_roc: 0.9671
Epoch 2/50
 - 103s - loss: 0.0424 - acc: 0.9840 - auc_roc: 0.9743 - val_loss: 0.0446 - val_acc: 0.9828 - val_auc_roc: 0.9781
Epoch 3/50
 - 103s - loss: 0.0362 - acc: 0.9861 - auc_roc: 0.9809 - val_loss: 0.0464 - val_acc: 0.9830 - val_auc_roc: 0.9824
Epoch 00003: early stopping
val loss: 0.04641580680064529
val acc: 0.9830121455699807
val auc roc: 0.9824134388407737

Train on 127657 samples, validate on 31914 samples
Epoch 1/50
 - 105s - loss: 0.0633 - acc: 0.9788 - auc_roc: 0.9283 - val_loss: 0.0457 - val_acc: 0.9827 - val_auc_roc: 0.9693
Epoch 2/50
 - 103s - loss: 0.0424 - acc: 0.9838 - auc_roc: 0.9756 - val_loss: 0.0441 - val_acc: 0.9835 - val_auc_roc: 0.9792
Epoch 3/50
 - 103s - loss: 0.0364 - acc: 0.9859 - auc_roc: 0.9817 - val_loss: 0.0462 - val_acc: 0.9831 - 

##### result

- d = 0.1

loss: 0.04655777110286902 (+/- 0.0007022790182632095)

acc: 0.9828101376355912 (+/- 0.00037164277486996864)

roc_auc: 0.9824663261111338 (+/- 0.0019216375163993913)


- d = 0.2

loss: 0.04468144988842106 (+/- 0.0007784231426185782)

acc: 0.9831161680493627 (+/- 0.0002944974584065599)

roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)


- d = 0.5

loss: 0.05021854950350052 (+/- 0.0032917323655443035)

acc: 0.9807723793697212 (+/- 0.0010786749474902403)

roc_auc: 0.974232999728601 (+/- 0.0029235142999876266)

In [18]:
eval_results(predictions)

key: fastText
loss: 0.04554556013987972 (+/- 0.0007409549985705233)
acc: 0.9828028274643259 (+/- 0.00036012916234944676)
roc_auc: 0.9803595673865427 (+/- 0.0005286042761330495)



key: glove
loss: 0.04468144988842106 (+/- 0.0007784231426185782)
acc: 0.9831161680493627 (+/- 0.0002944974584065599)
roc_auc: 0.9820037976582707 (+/- 0.0011390582860806907)



key: glove_twitter
loss: 0.04602299235323956 (+/- 0.0009691474923356518)
acc: 0.9830085934930389 (+/- 0.0005089609912555852)
roc_auc: 0.9803116532286843 (+/- 0.0028465203093875034)



key: 100
loss: 0.045756475632165755 (+/- 0.0011660776254660248)
acc: 0.9826607834588188 (+/- 0.00033719530397060746)
roc_auc: 0.9828798976335431 (+/- 0.0017559481870277032)



key: 300
loss: 0.04898223692728457 (+/- 0.0024399659468127824)
acc: 0.9822043513803225 (+/- 0.0006823144195393634)
roc_auc: 0.9806345902075753 (+/- 0.0016923229137481245)



key: 0.1
loss: 0.04655777110286902 (+/- 0.0007022790182632095)
acc: 0.9828101376355912 (+/- 0.0003716427748699

In [ ]:
opt

#### tuned params

In [ ]:
params_inception = {
    "maxlen": maxlen,
    "max_features": max_features,
    "embedding_size":,
    "embed_matrix":,
    "num_filter": 300,
    "reg": 0.0,
    "dropout": 0.2
}

### write result

In [7]:
write_prediction(opt, submission, "output/inception.csv")